web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,341 @@
+import { Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild, AbstractType} from '@angular/core';
+import { formatDate } from '@angular/common'
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+
+/**
+ * BarChartData includes all data used by the bar chart.
+ */
+interface BarChartData {
+  pushID: string; // Push ID number
+  state: number;  // Tag of push end state
+  startTime: any; // Start time of push, in `YYYY-MM-DD HH:mm` format
+  duration: number; // Duration of push, in hour
+}
+
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+type d3G = d3.Selection<SVGGElement, undefined, null, undefined>;
+type d3ScaleLinear = d3.ScaleLinear<number, number>;
+
+@Component({
+  selector: 'app-bar-chart',
+  templateUrl: './bar-chart.component.html',
+  styleUrls: ['./bar-chart.component.scss']
+})
+
+export class BarChartComponent implements OnChanges {
+  /**
+   * Private variables
+   * 
+   * The non-null assertion operator('!') suggests that there is no 
+   * issue of variables being `null` or `undefined`.
+   */
+  @ViewChild('barchart') private barChartContainer!: ElementRef; // Bar chart component
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() private currentPush!: step189_2020.IPushInfo | null;
+  private dataAll: BarChartData[] = [];
+  private dataComplete: BarChartData[] = [];
+  private totalDuration: number[] = [];
+  private completeDuration: number[] = [];
+  private svg!: d3SVG;
+  private focus!: d3G;
+  private brush!: d3G;
+  private brushSelector!: d3G;
+  private heightFocus!: number;
+  private heightBrush!: number;
+  private width!: number;
+  private xScaleFocus!: any;
+  private yScaleFocus!: d3ScaleLinear;
+  private xScaleBrush!: any;
+  private yScaleBrush!: d3ScaleLinear;
+  private xAxisFocus!: d3G;
+  private xAxisBrush!: d3G;
+  private yAxis!: d3G;
```

Comment:
(I tried doing this and yes, it works, just set the value to some initial state and remove the `!`, which should work just fine since you always reset it to another state anyway)